tests/test_extract.py

```diff
@@ -0,0 +1,146 @@
+                            'Not extracting file5.txt']
+        expected_absent = ['Not extracting file0_soft.txt',  # It's okay to extract the symlinks.
+                           'ERROR']
+        if use_hpss:
+            # It's okay to extract empty dirs.
+            expected_absent.append('Not extracting empty_dir')
+        self.check_strings(cmd, output + err, expected_present, expected_absent)
+
+        msg = 'Deleting the extracted files and doing it again, '
+        msg += 'while making sure the tars are kept.'
+        print(msg)
+        shutil.rmtree(self.test_dir)
+        os.mkdir(self.test_dir)
+        os.chdir(self.test_dir)
+        if not use_hpss:
+            shutil.copytree('{}/{}/{}'.format(TOP_LEVEL, self.backup_dir, self.cache), self.copy_dir)
+        cmd = '{}zstash extract -v --hpss={} --keep'.format(zstash_path, self.hpss_path)
+        output, err = run_cmd(cmd)
+        # Check that self.copy_dir contains all expected files
+        if not compare(os.listdir(self.copy_dir),
+                       ['index.db', '000000.tar', '000001.tar', '000002.tar', '000003.tar', '000004.tar']):
+            error_message = 'The zstash directory does not contain expected files.\nIt has: {}'.format(
+                os.listdir(self.copy_dir))
+            self.stop(error_message)
+        os.chdir(TOP_LEVEL)
+        expected_present = ['Extracting file0.txt', 'Extracting file0_hard.txt', 'Extracting file0_soft.txt',
+                            'Extracting file_empty.txt', 'Extracting dir/file1.txt', 'Extracting empty_dir',
+                            'Extracting dir2/file2.txt', 'Extracting file3.txt', 'Extracting file4.txt',
+                            'Extracting file5.txt']
+        if use_hpss:
+            expected_present.append('Transferring file from HPSS')
+        expected_absent = ['ERROR', 'Not extracting']
+        self.check_strings(cmd, output + err, expected_present, expected_absent)
+
+    def helperKeep(self, test_name, hpss_path, zstash_path=ZSTASH_PATH, keep=''):
+        """
+        Test `zstash extract` with `--keep` applied on `create`, `update`, or `extract`.
+        """
+        self.hpss_path = hpss_path
+        use_hpss = self.setupDirs(test_name)
+        self.create(use_hpss, zstash_path, keep=(keep=='create'))
+        self.add_files(use_hpss, zstash_path, keep=(keep=='update'))
+        self.extract(use_hpss, zstash_path)
+        msg = 'Deleting the extracted files and doing it again without verbose option, '
+        msg += 'while making sure the tars are kept.'
+        print_starred(msg)
+        self.assertWorkspace()
+        shutil.rmtree(self.test_dir)
+        os.mkdir(self.test_dir)
+        os.chdir(self.test_dir)
+        if not use_hpss:
+            shutil.copytree('{}/{}/{}'.format(TOP_LEVEL, self.backup_dir, self.cache), self.copy_dir)
+        #keep_option = ' --keep' if (keep == 'extract') else ''
+        keep_option = ' --keep'
```

Comment:
Using the second line makes `testKeepOnCreateHPSS` and `testKeepOnUpdateHPSS` pass.
- Why would those tests pass in the non-HPSS version using the first line?
- It looks like `--keep` has to be set on `extract` to actually do anything. Looking at the code though, it seems to me that `zstash create --keep` and `zstash update --keep` should update the `config` that is used in `zstash extract`. In fact, that's the only thing `--keep` on the first two command does. Only `zstash extract --keep` does anything noticeable (i.e., actually keeping the files). I would expect `zstash extract` to also keep files provided `zstash create --keep` or `zstash update --keep` is run first, but this doesn't appear to be the case in the HPSS version.